Sources/protoc-gen-swift/MessageGenerator.swift

```diff
@@ -498,12 +498,31 @@ class MessageGenerator {
   }
 }
 
-fileprivate func hasSingleMessageField(descriptor: Descriptor) -> Bool {
-  let result = descriptor.fields.contains {
-    // Repeated check also rules out maps.
-    ($0.type == .message || $0.type == .group) && $0.label != .repeated
+fileprivate func hasRecursiveSingularField(descriptor: Descriptor, visited: [Descriptor] = []) -> Bool {
+  var visited = visited
+  visited.append(descriptor)
+  return descriptor.fields.contains {
+    // Repeated fields use heap storage already
+    if ($0.type != .message && $0.type != .group) || $0.label == .repeated {
+      return false
+    }
+    guard let messageType = $0.messageType else {
+      return false
+    }
+
+    // We only care if the message or sub-message recurses to the root message.
+    if messageType === visited[0] {
+      return true
+    }
+
+    // Skip other visited fields.
+    if (visited[1...].contains { $0 === messageType }) {
```

Comment:
Actually, from a performance pov, you don't need this slice.  `visted[0]` was checked in the conditional above, so just check all of `visitited` and avoid making the temp object, etc.

Reply:
Ah, got it. I’m used to @golang where slicing is cheap and on the stack (shared storage with original slice).

Reply:
`visited[1...]` returns an `ArraySlice` that shares storage (and indices) with its original array, so it's the same in that regard.

Reply:
@allevato so is it correct to assume that sub slice doesn’t result in any additional heap allocations?

(I’m trying to understand if I should change it per @thomasvl’s suggestion).

Reply:
My thought was there is a call to make the slice (which may or may not have an alloc to its internal layout).  So that extra call up front (with potential allocs) is all to avoid one extra `===` check within the actually `contains(where:)` loop, so it didn't seem like trying to skip the first one was worth it compare to the cost of that one extra compare.

Reply:
Roger that. Updated in 1dafa90.